File: internal/store/outreach_drafts.go
package store

import (
	"database/sql"
	"fmt"
)

// Outreach draft statuses. Terminal: sent, failed, superseded. Active (at most
// one per posting): researching, awaiting_review, needs_work, no_hook.
const (
	DraftResearching    = "researching"
	DraftAwaitingReview = "awaiting_review"
	// DraftNeedsWork is a finished draft the doctrine judge rated below the
	// depth bar — reviewable/editable/sendable like awaiting_review, but
	// flagged so the user knows the judge wanted more.
	DraftNeedsWork = "needs_work"
	DraftNoHook    = "no_hook"
	DraftSent      = "sent"
	DraftFailed    = "failed"
	// DraftSuperseded is an awaiting_review/needs_work/no_hook draft retired by
	// a regenerate: kept in history (the user can still read it) but no longer
	// active, so a fresh draft can take its place.
	DraftSuperseded = "superseded"
)

// OutreachDraft is one pipeline run against a posting. Draft is what the
// pipeline assembled; Edited (when non-empty) is the user's revision and wins.
// Research/Hook/Lint/Violations/Critique carry stage outputs as JSON for the
// panel (Critique is the doctrine judge's verdict).
type OutreachDraft struct {
	ID        int64  `json:"id"`
	PostingID string `json:"posting_id"`
	Status    string `json:"status"`
	// Stage is the pipeline step an in-flight (researching) draft is on —
	// research|fill|humanize|honesty|judge — for the panel's progress bar. Empty
	// once the run reaches a terminal/review status.
	Stage      string `json:"stage"`
	Research   string `json:"research"`
	Hook       string `json:"hook"`
	Draft      string `json:"draft"`
	Edited     string `json:"edited"`
	Lint       string `json:"lint"`
	Violations string `json:"violations"`
	Critique   string `json:"critique"`
	FailReason string `json:"fail_reason"`
	CreatedAt  string `json:"created_at"`
	UpdatedAt  string `json:"updated_at"`
	SentAt     string `json:"sent_at,omitempty"`
}

const draftCols = `id, posting_id, status, stage, research, hook, draft, edited, lint,
violations, critique, fail_reason, created_at, updated_at, COALESCE(sent_at, '')`

func scanDraft(row interface{ Scan(...any) error }) (*OutreachDraft, error) {
	var d OutreachDraft
	err := row.Scan(&d.ID, &d.PostingID, &d.Status, &d.Stage, &d.Research, &d.Hook, &d.Draft,
		&d.Edited, &d.Lint, &d.Violations, &d.Critique, &d.FailReason, &d.CreatedAt, &d.UpdatedAt, &d.SentAt)
	if err != nil {
		return nil, err
	}
	return &d, nil
}

// CreateOutreachDraft starts a new draft for a posting. It refuses (with an
// error mentioning "active draft") when one is already in a non-terminal
// status — one review at a time per posting.
func (db *DB) CreateOutreachDraft(postingID string) (*OutreachDraft, error) {
	tx, err := db.Begin()
	if err != nil {
		return nil, err
	}
	defer tx.Rollback()

	var exists int
	if err := tx.QueryRow(`SELECT COUNT(1) FROM job_postings WHERE id = ?`, postingID).Scan(&exists); err != nil {
		return nil, err
	}
	if exists == 0 {
		return nil, sql.ErrNoRows
	}

	var active int
	err = tx.QueryRow(`SELECT COUNT(1) FROM outreach_drafts
WHERE posting_id = ? AND status IN (?, ?, ?, ?)`,
		postingID, DraftResearching, DraftAwaitingReview, DraftNeedsWork, DraftNoHook).Scan(&active)
	if err != nil {
		return nil, err
	}
	if active > 0 {
		return nil, fmt.Errorf("posting %s already has an active draft", postingID)
	}

	d, err := insertDraftTx(tx, postingID)
	if err != nil {
		return nil, err
	}
	return d, tx.Commit()
}

// RegenerateOutreachDraft retires the posting's current awaiting_review/
// needs_work/no_hook draft (→ superseded, kept in history) and starts a fresh
// one — the way to re-draft after backfilling experience/template/company info.
// It refuses while a draft is still researching (that run is pipeline-owned and
// in flight).
func (db *DB) RegenerateOutreachDraft(postingID string) (*OutreachDraft, error) {
	tx, err := db.Begin()
	if err != nil {
		return nil, err
	}
	defer tx.Rollback()

	var exists int
	if err := tx.QueryRow(`SELECT COUNT(1) FROM job_postings WHERE id = ?`, postingID).Scan(&exists); err != nil {
		return nil, err
	}
	if exists == 0 {
		return nil, sql.ErrNoRows
	}

	var researching int
	if err := tx.QueryRow(`SELECT COUNT(1) FROM outreach_drafts
WHERE posting_id = ? AND status = ?`, postingID, DraftResearching).Scan(&researching); err != nil {
		return nil, err
	}
	if researching > 0 {
		return nil, fmt.Errorf("posting %s already has an active draft", postingID)
	}

	// Carry the most recent research forward so a regenerate re-drafts against the
	// same web data instead of paying for another search.
	var priorResearch string
	_ = tx.QueryRow(`SELECT COALESCE(research, '') FROM outreach_drafts
WHERE posting_id = ? AND COALESCE(research, '') != '' ORDER BY id DESC LIMIT 1`, postingID).Scan(&priorResearch)

	if _, err := tx.Exec(`UPDATE outreach_drafts SET status = ?, updated_at = CURRENT_TIMESTAMP
WHERE posting_id = ? AND status IN (?, ?, ?)`,
		DraftSuperseded, postingID, DraftAwaitingReview, DraftNeedsWork, DraftNoHook); err != nil {
		return nil, err
	}

	d, err := insertDraftTx(tx, postingID)
	if err != nil {
		return nil, err
	}
	if priorResearch != "" {
		if _, err := tx.Exec(`UPDATE outreach_drafts SET research = ? WHERE id = ?`, priorResearch, d.ID); err != nil {
			return nil, err
		}
		d.Research = priorResearch
	}
	return d, tx.Commit()
}

// insertDraftTx inserts a fresh researching draft and returns it, within the
// caller's transaction.
func insertDraftTx(tx *sql.Tx, postingID string) (*OutreachDraft, error) {
	res, err := tx.Exec(`INSERT INTO outreach_drafts (posting_id) VALUES (?)`, postingID)
	if err != nil {
		return nil, err
	}
	id, err := res.LastInsertId()
	if err != nil {
		return nil, err
	}
	return scanDraft(tx.QueryRow(`SELECT `+draftCols+` FROM outreach_drafts WHERE id = ?`, id))
}

// GetOutreachDraft returns one draft, or (nil, nil) when absent.
func (db *DB) GetOutreachDraft(id int64) (*OutreachDraft, error) {
	d, err := scanDraft(db.QueryRow(`SELECT `+draftCols+` FROM outreach_drafts WHERE id = ?`, id))
	if err == sql.ErrNoRows {
		return nil, nil
	}
	return d, err
}

// ListOutreachDrafts returns a posting's drafts, newest first.
func (db *DB) ListOutreachDrafts(postingID string) ([]OutreachDraft, error) {
	rows, err := db.Query(`SELECT `+draftCols+` FROM outreach_drafts
WHERE posting_id = ? ORDER BY id DESC`, postingID)
	if err != nil {
		return nil, err
	}
	defer rows.Close()
	var out []OutreachDraft
	for rows.Next() {
		d, err := scanDraft(rows)
		if err != nil {
			return nil, err
		}
		out = append(out, *d)
	}
	return out, rows.Err()
}

// SetOutreachDraftStage records which pipeline step an in-flight draft is on, so
// the panel's progress bar can advance as the run proceeds. Best-effort from the
// engine's view (a failed stage write must not abort the run), and a no-op-safe
// update: it never changes status, only the stage marker.
func (db *DB) SetOutreachDraftStage(id int64, stage string) error {
	res, err := db.Exec(`UPDATE outreach_drafts SET
stage = ?, updated_at = CURRENT_TIMESTAMP WHERE id = ?`, stage, id)
	if err != nil {
		return err
	}
	return mustAffect(res)
}

// SetOutreachDraftResult records a pipeline outcome: the new status plus any
// stage outputs. Empty strings overwrite (stages own their fields). Reaching a
// result clears the in-flight stage marker — the progress bar only shows while
// researching.
func (db *DB) SetOutreachDraftResult(id int64, status, research, hook, draft, lint, violations, critique, failReason string) error {
	res, err := db.Exec(`UPDATE outreach_drafts SET
status = ?, stage = '', research = ?, hook = ?, draft = ?, lint = ?, violations = ?,
critique = ?, fail_reason = ?, updated_at = CURRENT_TIMESTAMP WHERE id = ?`,
		status, research, hook, draft, lint, violations, critique, failReason, id)
	if err != nil {
		return err
	}
	return mustAffect(res)
}

// SetOutreachDraftEdited stores the user's revision and its lint findings.
func (db *DB) SetOutreachDraftEdited(id int64, edited, lint string) error {
	res, err := db.Exec(`UPDATE outreach_drafts SET
edited = ?, lint = ?, updated_at = CURRENT_TIMESTAMP WHERE id = ?`, edited, lint, id)
	if err != nil {
		return err
	}
	return mustAffect(res)
}

// MarkOutreachDraftSent flips a draft to sent and, in the same transaction,
// bumps the posting's outreach_count and stamps last_outreach_at — the send
// date is what makes Touch-2 follow-ups cheap later.
func (db *DB) MarkOutreachDraftSent(id int64) (*OutreachDraft, error) {
	tx, err := db.Begin()
	if err != nil {
		return nil, err
	}
	defer tx.Rollback()

	res, err := tx.Exec(`UPDATE outreach_drafts SET
status = ?, sent_at = CURRENT_TIMESTAMP, updated_at = CURRENT_TIMESTAMP
WHERE id = ? AND status != ?`, DraftSent, id, DraftSent)
	if err != nil {
		return nil, err
	}
	if err := mustAffect(res); err != nil {
		// Idempotent double-send: an already-sent draft returns itself (the
		// panel's Mark-sent can be double-clicked); only a missing row errors.
		if d, gErr := scanDraft(tx.QueryRow(`SELECT `+draftCols+` FROM outreach_drafts WHERE id = ?`, id)); gErr == nil && d.Status == DraftSent {
			return d, tx.Commit()
		}
		return nil, err
	}
	// The send completes the "next up" to-do, so the queue mark clears too.
	if _, err := tx.Exec(`UPDATE job_postings SET
outreach_count = outreach_count + 1, last_outreach_at = DATE('now'), next_up_at = NULL
WHERE id = (SELECT posting_id FROM outreach_drafts WHERE id = ?)`, id); err != nil {
		return nil, err
	}
	d, err := scanDraft(tx.QueryRow(`SELECT `+draftCols+` FROM outreach_drafts WHERE id = ?`, id))
	if err != nil {
		return nil, err
	}
	return d, tx.Commit()
}

func mustAffect(res sql.Result) error {
	n, err := res.RowsAffected()
	if err != nil {
		return err
	}
	if n == 0 {
		return sql.ErrNoRows
	}
	return nil
}

// ReapStuckOutreachDrafts fails any draft stuck in `researching` longer than
// olderThanMinutes (0 = all of them). A row only stays in researching while an
// engine goroutine is live; after a server restart (or a crash the catch-all
// couldn't see) the row is orphaned — it blocks new drafts for its posting and
// the panel polls it forever. Called at serve startup, before any new work.
func (db *DB) ReapStuckOutreachDrafts(olderThanMinutes int) (int64, error) {
	res, err := db.Exec(`UPDATE outreach_drafts SET
status = ?, fail_reason = 'interrupted — scout restarted mid-run',
updated_at = CURRENT_TIMESTAMP
WHERE status = ? AND updated_at <= datetime('now', ?)`,
		DraftFailed, DraftResearching, fmt.Sprintf("-%d minutes", olderThanMinutes))
	if err != nil {
		return 0, err
	}
	return res.RowsAffected()
}

File: scout/store/outreach_drafts.py
"""Outreach draft pipeline rows. Port of internal/store/outreach_drafts.go."""
from __future__ import annotations

import sqlite3
from dataclasses import dataclass

from . import errors
from ._helpers import tx

# Draft statuses. Terminal: sent, failed, superseded.
# Active (at most one per posting): researching, awaiting_review, needs_work, no_hook.
DRAFT_RESEARCHING = "researching"
DRAFT_AWAITING_REVIEW = "awaiting_review"
DRAFT_NEEDS_WORK = "needs_work"
DRAFT_NO_HOOK = "no_hook"
DRAFT_SENT = "sent"
DRAFT_FAILED = "failed"
DRAFT_SUPERSEDED = "superseded"


@dataclass
class OutreachDraft:
    id: int = 0
    posting_id: str = ""
    status: str = ""
    stage: str = ""
    research: str = ""
    hook: str = ""
    draft: str = ""
    edited: str = ""
    lint: str = ""
    violations: str = ""
    critique: str = ""
    fail_reason: str = ""
    created_at: str = ""
    updated_at: str = ""
    sent_at: str = ""


_DRAFT_COLS = (
    "id, posting_id, status, stage, research, hook, draft, edited, lint, "
    "violations, critique, fail_reason, created_at, updated_at, COALESCE(sent_at, '')"
)


def _scan_draft(row) -> OutreachDraft:
    return OutreachDraft(
        id=row[0], posting_id=row[1], status=row[2], stage=row[3], research=row[4],
        hook=row[5], draft=row[6], edited=row[7], lint=row[8], violations=row[9],
        critique=row[10], fail_reason=row[11], created_at=row[12], updated_at=row[13], sent_at=row[14],
    )


def _must_affect(cur: sqlite3.Cursor) -> None:
    if cur.rowcount == 0:
        raise errors.NotFound()


def _insert_draft(con: sqlite3.Connection, posting_id: str) -> OutreachDraft:
    """Insert a fresh researching draft and return it (within the caller's tx)."""
    cur = con.execute("INSERT INTO outreach_drafts (posting_id) VALUES (?)", (posting_id,))
    row = con.execute(f"SELECT {_DRAFT_COLS} FROM outreach_drafts WHERE id = ?", (cur.lastrowid,)).fetchone()
    return _scan_draft(row)


def create_outreach_draft(con: sqlite3.Connection, posting_id: str) -> OutreachDraft:
    """Start a new draft for a posting. Raises NotFound for an unknown posting and
    ValueError ("active draft") when one is already in a non-terminal status."""
    with tx(con):
        if con.execute("SELECT COUNT(1) FROM job_postings WHERE id = ?", (posting_id,)).fetchone()[0] == 0:
            raise errors.NotFound()
        active = con.execute(
            "SELECT COUNT(1) FROM outreach_drafts WHERE posting_id = ? AND status IN (?, ?, ?, ?)",
            (posting_id, DRAFT_RESEARCHING, DRAFT_AWAITING_REVIEW, DRAFT_NEEDS_WORK, DRAFT_NO_HOOK),
        ).fetchone()[0]
        if active > 0:
            raise ValueError(f"posting {posting_id} already has an active draft")
        d = _insert_draft(con, posting_id)
    return d


def regenerate_outreach_draft(con: sqlite3.Connection, posting_id: str) -> OutreachDraft:
    """Retire the posting's current reviewable draft (→ superseded) and start a
    fresh one, carrying the most recent research forward. Refuses while a draft is
    still researching."""
    with tx(con):
        if con.execute("SELECT COUNT(1) FROM job_postings WHERE id = ?", (posting_id,)).fetchone()[0] == 0:
            raise errors.NotFound()
        researching = con.execute(
            "SELECT COUNT(1) FROM outreach_drafts WHERE posting_id = ? AND status = ?",
            (posting_id, DRAFT_RESEARCHING),
        ).fetchone()[0]
        if researching > 0:
            raise ValueError(f"posting {posting_id} already has an active draft")

        prior_row = con.execute(
            "SELECT COALESCE(research, '') FROM outreach_drafts "
            "WHERE posting_id = ? AND COALESCE(research, '') != '' ORDER BY id DESC LIMIT 1",
            (posting_id,),
        ).fetchone()
        prior_research = prior_row[0] if prior_row is not None else ""

        con.execute(
            "UPDATE outreach_drafts SET status = ?, updated_at = CURRENT_TIMESTAMP "
            "WHERE posting_id = ? AND status IN (?, ?, ?)",
            (DRAFT_SUPERSEDED, posting_id, DRAFT_AWAITING_REVIEW, DRAFT_NEEDS_WORK, DRAFT_NO_HOOK),
        )

        d = _insert_draft(con, posting_id)
        if prior_research != "":
            con.execute("UPDATE outreach_drafts SET research = ? WHERE id = ?", (prior_research, d.id))
            d.research = prior_research
    return d


def get_outreach_draft(con: sqlite3.Connection, id: int) -> OutreachDraft | None:
    """Return one draft, or None when absent."""
    row = con.execute(f"SELECT {_DRAFT_COLS} FROM outreach_drafts WHERE id = ?", (id,)).fetchone()
    return _scan_draft(row) if row is not None else None


def list_outreach_drafts(con: sqlite3.Connection, posting_id: str) -> list[OutreachDraft]:
    """A posting's drafts, newest first."""
    rows = con.execute(
        f"SELECT {_DRAFT_COLS} FROM outreach_drafts WHERE posting_id = ? ORDER BY id DESC",
        (posting_id,),
    ).fetchall()
    return [_scan_draft(r) for r in rows]


def set_outreach_draft_stage(con: sqlite3.Connection, id: int, stage: str) -> None:
    """Record which pipeline step an in-flight draft is on."""
    cur = con.execute(
        "UPDATE outreach_drafts SET stage = ?, updated_at = CURRENT_TIMESTAMP WHERE id = ?", (stage, id)
    )
    _must_affect(cur)


def set_outreach_draft_result(
    con: sqlite3.Connection, id: int, status: str, research: str, hook: str,
    draft: str, lint: str, violations: str, critique: str, fail_reason: str,
) -> None:
    """Record a pipeline outcome: the new status plus any stage outputs."""
    cur = con.execute(
        "UPDATE outreach_drafts SET "
        "status = ?, stage = '', research = ?, hook = ?, draft = ?, lint = ?, violations = ?, "
        "critique = ?, fail_reason = ?, updated_at = CURRENT_TIMESTAMP WHERE id = ?",
        (status, research, hook, draft, lint, violations, critique, fail_reason, id),
    )
    _must_affect(cur)


def set_outreach_draft_edited(con: sqlite3.Connection, id: int, edited: str, lint: str) -> None:
    """Store the user's revision and its lint findings."""
    cur = con.execute(
        "UPDATE outreach_drafts SET edited = ?, lint = ?, updated_at = CURRENT_TIMESTAMP WHERE id = ?",
        (edited, lint, id),
    )
    _must_affect(cur)


def mark_outreach_draft_sent(con: sqlite3.Connection, id: int) -> OutreachDraft:
    """Flip a draft to sent and stamp sent_at. Idempotent: an already-sent draft
    returns itself; only a missing row raises NotFound."""
    cur = con.execute(
        "UPDATE outreach_drafts SET status = ?, sent_at = CURRENT_TIMESTAMP, updated_at = CURRENT_TIMESTAMP "
        "WHERE id = ? AND status != ?",
        (DRAFT_SENT, id, DRAFT_SENT),
    )
    if cur.rowcount == 0:
        row = con.execute(f"SELECT {_DRAFT_COLS} FROM outreach_drafts WHERE id = ?", (id,)).fetchone()
        if row is not None:
            d = _scan_draft(row)
            if d.status == DRAFT_SENT:
                return d
        raise errors.NotFound()
    row = con.execute(f"SELECT {_DRAFT_COLS} FROM outreach_drafts WHERE id = ?", (id,)).fetchone()
    return _scan_draft(row)


def reap_stuck_outreach_drafts(con: sqlite3.Connection, older_than_minutes: int) -> int:
    """Fail any draft stuck in `researching` longer than older_than_minutes
    (0 = all). Returns the number reaped."""
    cur = con.execute(
        "UPDATE outreach_drafts SET "
        "status = ?, fail_reason = 'interrupted — scout restarted mid-run', updated_at = CURRENT_TIMESTAMP "
        "WHERE status = ? AND updated_at <= datetime('now', ?)",
        (DRAFT_FAILED, DRAFT_RESEARCHING, f"-{older_than_minutes} minutes"),
    )
    return cur.rowcount
